accepted/future-releases/records/records-feature-specification.md

```diff
@@ -252,10 +252,100 @@ ambiguity, we disambiguate by treating `on` as a clause for `try` and not a
+variable declaration may omit a preceding type annotation. Likewise, a function
+declaration may omit a preceding return type. This combination of syntax where
+an optional trailing element is followed by syntax with an optional preceding
+element can lead to ambiguity. In particular:
+
+```dart
+@metadata (a, b) function() {}
+```
+
+This could be a metadata annotation `@metadata(a, b)` associated with a function
+declaration with no return type. Or it could be a metadata annotation
+`@metadata` associated with a function whose return type is the record type `(a,
+b)`.
+
+In practice, idiomatically written code is clear thanks to whitespace:
+
+```dart
+@metadata(a, b) function() {}
+
+@metadata (a, b) function() {}
+```
+
+The former applies `(a, b)` to the metadata annotation and the latter is a
+return type. We disambiguate in the same way, by making whitespace after a
+metadata annotation name significant. Change the grammar to:
+
+```
+metadatum ::= identifier                                    // Existing rule.
+            | qualifiedName                                 // Existing rule.
+            | constructorDesignation NO_SPACE arguments  // Changed.
+```
+
+The `NO_SPACE` lexical rule matches when there are no whitespace characters or
+comments (according to the existing `WHITESPACE` and `COMMENT` lexical rules)
+between the `constructorDesignation` and `arguments`. In other words, for an
+argument list to be part of the metadata annotation, the `(` must occur
+immediately after the last character in the `constructorDesignation`. The last
+character in `constructorDesignation` may be an identifier or the `>` in a
+type argument list.
+
+```dart
+// These are parsed as argument lists to the annotation:
+@metadata(x, y) a;
+
+@metadata<T>(x, y) a;
+
+@metadata <T>(x, y) a;
+
+// These are parsed as record variable types:
+@metadata (x, y) a;
+
+@metadata
+(x, y) a;
+
+@metadata/* comment */(x, y) a;
+
+@metadata // Comment.
+(x,) a;
+```
+
+Note that the `NO_SPACE` rule is applied unconditionally, even when the metadata
+annotation appears in a context where no ambiguity with record types is
+possible, as in:
+
+```dart
+@metadata (x, y)
+class C {}
+```
+
+This example has a syntax error because the `(x, y)` is not parsed as arguments
+to the metadata and can't be parsed as anything else either.
+
+Another interesting case is:
+
+```dart
+@metadata<T> (x, y) a;
+```
+
+This is a syntax error because the `<T>` means there *must* be an argument list
+after it, but the `NO_SPACE` in `metadatum` prevents it from being parsed as
+such and the result is an error. *We could ignore whitespace after the `>` by
+tweaking the grammar, but we choose to require `NO_SPACE` even here since
+`@metadata<T>` appears to be a generic instantiation and could potentially be
+valid syntax in the future.*
+
+**Breaking change:** Existing metadata annotations with whitespace before their
```

Comment:
We should clarify whether this is a hard breaking change or language versioned.  @jensjoha I would imagine you would prefer not to language version this?

Reply:
Yes.

Reply:
Would we want the `on`-after-`try`-`catch` to be a breaking change too?

Reply:
That's what was suggested in https://github.com/dart-lang/language/issues/2406, right?

Reply:
Not sure I was being that precise, I was mainly thinking of whether code needed to be migrated, not whether it should be fixed. But since there likely are no code to fix/migrate, I'm fine with either choice.

Reply:
I kind of assumed we would gate this behind a language version since the code it disambiguates can only appear in a later language version too. Is there a reason we wouldn't?

Reply:
For one thing the parser doesn't know about language versioning.

Reply:
Also if the parser _was_ to know about language versioning (and for instance not know about a feature that's not turned on), recovery is going to suffer.

Reply:
Have we really not made any breaking changes since 2.7? (Like we did with `f(a<b,c>(d))` in Dart 2.0.)

I don't think we can promise to not make such a breaking change in the future.

Parsing incremental features even in libraries with a language-version which doesn't support the feature, is fine, but if we have breaking changes, where the *same* syntax parses *differently* in a later version, we should be able to support that too. That means finding the language version before parsing the code, and parsing differently depending on the language version.

We designed the `//@dart=2.7` syntax, and placement of it, so that it should always be possible to read the language marker before any actual Dart code. As long as we don't change comment syntax based on language version, the parser can see the language version marker first. (It will have to look for it, obviously.)

Reply:
OK, updated to clarify that the disambiguation is not gated on the language version.

Reply:
The "parser" is split into two: the scanner and the parser.
Currently the scanner has some booleans about features, e.g. extension types which affect whether "extension" is marked as a keyword or an identifier. (to be honest I'm not sure that's necessary, but that's how it is) whereas the parser knows nothing (*).
The parser gets a token stream and then parses the language as it knows it. This also means that when getting an invalid program it can recover - try to make sense of it - with the full language. For instance if it sees `(a, b)` it will report it to the listener as a record (type or literal depending on where it is) and the listener will then say "you don't have records enabled. do this to enable it" if records aren't enabled.
If the parser instead was to recover without records it would spew lots of errors about "unexpected comma" or "expected end-paren" or whatever, which is a lot less useful imo.

Now, I understand that if it's the same syntax meaning different things, recovery (maybe) wouldn't suffer. I can't really tell what's worse, though, code that silently means something different when I upgrade my sdk, or code that silently means something different when I upgrade my min sdk version in my pubspec.yaml... Both are pretty bad (but I was also a fan of the `Record(a, b)` syntax that didn't have these problems).

*: because of the booleans in the scanner, triple shift will become two tokens when triple shift is not enabled and we had to hack better recovery around that in the parser which wasn't great (in that recovery we, as I recall, give an error message about language versions although we're really guessing).

Reply:
Actually, in the case around `on` I probably do know which is worse.
If it's unconditionally the new thing and one upgrades the sdk but not the pubspec one will get an error saying you don't have records enabled.
If it's conditionally on the language version, upgrading won't give an error. Then changing the language version won't give an error, but it will mean something different.

Reply:
OK, so there's two pieces here:

* How the parser handles record expressions in libraries whose language version doesn't support records. I think it makes perfect sense to parse that as a record expression and then report a graceful error about not having the feature enabled. The syntax isn't valid anyway, so the parser's going to have some kind of error either way and the most likely thing the user thought they were doing was writing a record.

* How the parser handles the way we disambiguated code to deal with records. Concretely, that means an `on` followed by a parenthesized thing after a `try` block and whitespace between a metadata annotation name and the parameter list. If the parser ignores language versioning and applies the new disambiguation rules even to old code, then code which used to be fine now has an error.

The latter feels unnecessarily breaking to me and normally I'd suggest we not do that. But given that this seems to not really break any actual code in the wild, I think it's probably fine too. That's what the latest version of this PR specifies. Is that OK?

Reply:
I am certainly fine with it not being guarded by language versions --- that much was probably clear so the question was likely meant for someone else :)